plugins/backstage-plugin-coder/src/api/Auth.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * @file Defines shared values and types used among any custom Coder auth
+ * implementations for the frontend.
+ */
+import { createApiRef } from '@backstage/core-plugin-api';
+import { CODER_API_REF_ID_PREFIX } from '../typesConstants';
+
+type AuthSubscriptionPayload = Readonly<{
+  token: string;
+  isTokenValid: boolean;
+}>;
+
+export type AuthSubscriptionCallback<
+  TSubscriptionPayload extends AuthSubscriptionPayload = AuthSubscriptionPayload,
+> = (payload: TSubscriptionPayload) => void;
+
+export type AuthValidatorDispatch = (newStatus: boolean) => void;
+
+/**
+ * Shared set of properties among all Coder auth implementations
+ */
+export type CoderAuthApi<
+  TPayload extends AuthSubscriptionPayload = AuthSubscriptionPayload,
+> = TPayload & {
+  /**
+   * Gives back a "state setter" that lets a different class dispatch a new auth
+   * status to the auth class implementation.
+   *
+   * Use this to send the new status you think the auth should have. The auth
+   * will decide whether it will let the dispatch go through and update state.
+   */
+  getAuthStateSetter: () => AuthValidatorDispatch;
```

Comment:
I think this part is a little unconventional (it's kind of bringing React's `useState` into classes), but after playing around with a few different options, it felt like the easiest to use, and the easiest to maintain, while keeping the class completely decoupled from the API logic

I originally had something like this:
```ts
type GetAuthValueCallback = () => boolean | Promise<boolean>;

type CoderAuthApi = {
  setAuth: (callback: GetAuthValueCallback) => void;
};
```
But in practice, it led to weird, hard-to-reason-about behavior, because you had to worry about things like:
- Nested try/catches in async logic, and figuring out how that all interacted with the JS event loop
- React Query needs you to throw an error to indicate that something failed, so you had to throw an error through multiple layers of callbacks
- Lot of extra indentation in general

I don't have the PR open for it open yet, but here's the method from the `CoderClient`, which I think is easier to reason about, because it mostly reads top-to-bottom:
```ts
validateAuth = async (): Promise<boolean> => {
    const dispatchNewStatus = this.authApi.getAuthStateSetter();

    try {
      // Dummy request; just need something that all users would have access
      // to, and that doesn't require a body
      await this.sdkApi.getUserLoginType();
      dispatchNewStatus(true);
      return true;
    } catch (err) {
      dispatchNewStatus(false);

      if (!(err instanceof AxiosError)) {
        throw err;
      }

      const response = err.response;
      if (response === undefined) {
        throw new Error('Unable to complete request - unknown error detected.');
      }

      if (response.status >= 400 && response.status !== 401) {
        throw new BackstageHttpError('Failed to complete request', response);
      }
    }

    return false;
  };
```

Reply:
Seems chill to me.